### DISCUSSION TITLE
Ffmpeg usage VS Frigate FFmpeg usage

### DISCUSSION THREAD

**Discussion by anthurta** [blakeblackshear/frigate#6772] [Ask A Question] [answered]
Hello,

First of all, congratulations in this development. It is really nice the work you have done.

I am trying to setup a simple NVR with some Dahua cameras. I just want Frigate to provide the recording function continuously: no detection of objects/person or motion needed, as the detection (motion and intelligent tracking function is integrated in the cameras already (I can get events from Cameras when detection is triggered, for instance).

For this, I am struggling to set up them because the CPU usage. The cameras are already in h265 codec. I don't want to encode/transcode, just to record and stores the files in whatever format, just for some days in case that I need to watch them (that 99.9% will not be the case and they will be just deleted).

I am running Frigate in LXC through Proxmox

Having said that, I have compared a simple FFmpeg command CPU usage VS CPU usage from Frigate (with a configuration, I hope, to just record):

`ffmpeg -hide_banner -y -rtsp_transport tcp -use_wallclock_as_timestamps 1 -i "rtsp://user:pass@192.168.0.106/cam/realmonitor?channel=1&subtype=0" -vcodec copy -acodec copy -f segment -reset_timestamps 1 -segment_time 30 -segment_format mkv -segment_atclocktime 1 -strftime 1 %Y%m%dT%H%M%S.mkv`

Ffmpeg output from Command line:

`Guessed Channel Layout for Input Stream #0.1 : mono
Input #0, rtsp, from 'rtsp://user:pass@192.168.0.106/cam/realmonitor?channel=1&subtype=0':
  Metadata:
    title           : Media Server
  Duration: N/A, start: 1686558485.581750, bitrate: N/A
    Stream #0:0: Video: hevc (Main), yuvj420p(pc, bt470bg/bt470bg/smpte170m), 3840x2160, 20 fps, 20 tbr, 90k tbn, 20 tbc
    Stream #0:1: Audio: pcm_alaw, 8000 Hz, mono, s16, 64 kb/s
[segment @ 0x5595c0363440] Opening '20230612T082806.mkv' for writing
Output #0, segment, to '%Y%m%dT%H%M%S.mkv':
  Metadata:
    title           : Media Server
    encoder         : Lavf58.45.100
    Stream #0:0: Video: hevc (Main), yuvj420p(pc, bt470bg/bt470bg/smpte170m), 3840x2160, q=2-31, 20 fps, 20 tbr, 1k tbn, 20 tbc
    Stream #0:1: Audio: pcm_alaw, 8000 Hz, mono, s16, 64 kb/s
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #0:1 -> #0:1 (copy)
Press [q] to stop, [?] for help`

CPU Usage:

> top - 10:30:09 up  2:42,  1 user,  load average: 0.83, 1.74, 1.94
> Tasks: 233 total,   1 running, 227 sleeping,   5 stopped,   0 zombie
> %Cpu(s): 22.2 us,  6.1 sy,  0.0 ni, 71.2 id,  0.2 wa,  0.0 hi,  0.3 si,  0.0 st
> MiB Mem :  15685.3 total,   7138.3 free,   6571.2 used,   1975.8 buff/cache
> MiB Swap:   7168.0 total,   7168.0 free,      0.0 used.   8759.0 avail Mem 
>     PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                                           
>    2717 root      20   0 7669784   5.0g  18240 S  98.0  32.5 166:15.01 kvm                                                                                                                                                                               
>   38035 root      20   0  234604  49132  35672 S   1.3   0.3   0:01.59 ffmpeg                                                                                                                                                                            
>    2758 root      20   0       0      0      0 S   1.0   0.0   2:08.45 vhost-2717                                                                                                                                                                        
>     412 root      20   0   80432  33408  10860 S   0.3   0.2   0:00.37 dmeventd                                                                                                                                                                          
>     807 jellyfin  20   0 4546748 177632  82544 S   0.3   1.1   0:15.40 jellyfin                                                                                                                                                                          
>    1099 www-data  20   0  369812 147148  11660 S   0.3   0.9   0:03.51 pveproxy worker                                                                                                                                                                   
>    1100 www-data  20   0  367072 145792  12656 S   0.3   0.9   0:03.30 pveproxy worker                                                                                                                                                                   
>       1 root      20   0  164412  10648   7728 S   0.0   0.1   0:00.62 systemd                                                                                                                                                                           
>       2 root      20   0       0      0      0 S   0.0   0.0   0:00.00 kthreadd                                                                                                                                                                          
>       3 root       0 -20       0      0      0 I   0.0   0.0   0:00.00 rcu_gp

So, very low cpu usage (1.3%)

When I go to Frigate, the story changes. This is my config file:

> mqtt:
>   enabled: False
>   host: 192.168.0.30
>   port: 1883
>   user: homeassistant
>   password: pass
> birdseye:
>   enabled: False
>   mode: continuous
> 
> logger:
>   default: debug
> 
> detect:
>   enabled: False
>   width: 3840
>   height: 2160
> 
>   
> cameras:
>   SPCamera: # <------ Name the camera
>     ffmpeg:
>       inputs:
>         - path: rtsp://user:pass@192.168.0.106/cam/realmonitor?channel=1&subtype=0 # <----- The stream you want to use for detection
>           roles:
>             - record
>             - detect
>     record:
>       enabled: True
>       retain:
>         days: 30
>         mode: all

And this is the CPU usage:

> top - 10:34:05 up  2:46,  1 user,  load average: 2.29, 1.73, 1.86

> Tasks: 267 total,   2 running, 265 sleeping,   0 stopped,   0 zombie
> %Cpu(s): 55.5 us,  2.7 sy,  0.0 ni, 40.8 id,  0.1 wa,  0.0 hi,  0.8 si,  0.0 st
> MiB Mem :  15685.3 total,   6603.2 free,   6983.8 used,   2098.2 buff/cache
> MiB Swap:   7168.0 total,   7168.0 free,      0.0 used.   8324.4 avail Mem 
>     PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND                                                                                                                                                                           
>    2717 root      20   0 7669784   5.0g  18240 S 149.7  32.5 170:40.09 kvm                                                                                                                                                                               
>   40056 root      20   0  531380 151212  17896 R  77.5   0.9   0:23.84 ffmpeg                                                                                                                                                                            
>   40050 root      20   0  952924 114560   7360 S   5.6   0.7   0:01.85 frigate.capture                                                                                                                                                                   
>    2758 root      20   0       0      0      0 S   1.7   0.0   2:11.64 vhost-2717                                                                                                                                                                        
>   40046 root      20   0  760604 109956  12192 S   0.7   0.7   0:00.39 frigate.output                                                                                                                                                                    
>   21593 root      20   0       0      0      0 I   0.3   0.0   0:00.48 kworker/u8:2-flush-253:1                                                                                                                                                          
>   32686 www-data  20   0  365564 142876  11244 S   0.3   0.9   0:01.50 pveproxy worker                                                                                                                                                                   
>   39871 root      20   0 1566836 192708  57996 S   0.3   1.2   0:01.36 python3      

Here, the FFMpeg usage (from Frigate) is much higher, at 77.5%

So I am pretty sure I am doing some dumb  mistake in the configuration, but I am not able to find it.

Can anyone help me? To reitarete: I just want to record continuously with no detection enabled, just "copy/paste" the stream from the camera in the coded format.

Thanks,
Antonio

**Comment by blakeblackshear** [comment, marked as answer]
Frigate requires that at least one stream be decoded from the camera. It cannot currently be used for recording alone with no other functions.

**Reply by anthurta** [reply]
Thanks for the answer. Is there any plans to avoid decoding? Or is this a really strange use case? I am thinking that nowadays more and more cameras come with some kind of intelligent detection that may reduce the need of having SW implementation.

At the end, I added two streams per camera:
- 4K stream that is the one I am recording
- CIF stream that is the one needed from Frigate to be decoded so the CPU usage is very low (using just 1 core). I have just selected the lower resolution stream possible for my camera, with a low quality and just 1 FPS (it can't be poorer!)

Pasting here my config file in case is helpful for anyone, although the trick is really easy:

detect:
  enabled: False
  width: 3840
  height: 2160

record:
      enabled: True
      retain:
        days: 30
        mode: all

  
cameras:
  SPCamera: # <------ Name the camera
    ffmpeg:
      inputs:
        - path: rtsp://us:pss@192.168.0.105/cam/realmonitor?channel=1&subtype=1 # The stream for detection
          roles:
            - detect
        - path: rtsp://us:pss@192.168.0.106/cam/realmonitor?channel=1&subtype=0 # <----- The stream for recording
          roles:
            - record

**Reply by NickM-27** [reply]
This is still a suboptimal setup, when you set the width / height to 4K resolution it is scaling up the decoded stream to 4k. You'd be better off setting the detect width and height to the smaller actual size of the stream

**Reply by anthurta** [reply]
Exactly, very good point. Now the CPU usage is even lower!
